src/centos/macros.tpl

```diff
@@ -41,6 +41,19 @@ RUN \
 {% else %}
     virtualenv-$PYTHON_VERSION ${APP_ROOT} && \
 {% endif %}
+{% if spec.version != "2.7" %}
+    # Python 3 only code, Python 2 installs pip from PyPI in the assemble script. \
+    # We have to upgrade pip to a newer verison because: \
+    # * pip < 9 does not support different packages' versions for Python 2/3 \
+    # * pip < 19.3 does not support manylinux2014 wheels. Only manylinux2014 (and later) wheels \
+    #   support platforms like ppc64le, aarch64 or armv7 \
+    # We are newly using wheel from one of the latest stable Fedora releases (from RPM python-pip-wheel) \
+    # because it's tested better that whatever version from PyPI and contains useful patches. \
+    # We have to do it here so the permissions are correctly fixed and pip is able  to reinstall \
+    # itself in the next build phases in the assemble script if user wants the latest version \
```

Comment:
>   We have to do it here so the permissions are correctly fixed and pip is able  to reinstall
>   itself in the next build phases in the assemble script if user wants the latest version

I don't understand the reasoning here. Would there really be some problem with permissions if we installed the wheel in the assemble script? I don't know of any (but I might be wrong).

I think we're installing it here during the build because it's better to have it already installed in the image — assemble script will run faster, and we can remove the original wheels directory from the image before shipping it, thus it's lighter. How about:

```
We install the wheel here during the build of the image so that the assembly script runs faster
and so that we can remove the wheel from the image after installing it.
```

Reply:
It's not about Dockerfile vs assemble script, it's a comment describing why the installation is a part of the macro. The fact that the installation has to happen during podman build is not questionable. But the installation has to happen between the step when venv is created and the step when we set the proper permissions for the reasons described in the comment. And because the two steps are currently in this macro, the installation step has to be there as well or we have to split the macro into multiple macros.